magicblock-task-scheduler/tests/service.rs

```diff
@@ -0,0 +1,213 @@
+    let (env, token, handle) = setup()?;
+
+    let account =
+        env.create_account_with_config(LAMPORTS_PER_SOL, 1, guinea::ID);
+
+    // Schedule a task
+    let ix = Instruction::new_with_bincode(
+        guinea::ID,
+        &GuineaInstruction::ScheduleTask(ScheduleTaskArgs {
+            task_id: 1,
+            execution_interval_millis: 10,
+            iterations: 1,
+            instructions: vec![Instruction::new_with_bincode(
+                guinea::ID,
+                &GuineaInstruction::Increment,
+                vec![AccountMeta::new(account.pubkey(), false)],
+            )],
+        }),
+        vec![
+            AccountMeta::new_readonly(magicblock_magic_program_api::ID, false),
+            AccountMeta::new(env.payer.pubkey(), true),
+            AccountMeta::new(account.pubkey(), false),
+        ],
+    );
+    let txn = env.build_transaction(&[ix]);
+    let result = env.execute_transaction(txn).await;
+    assert!(
+        result.is_ok(),
+        "failed to execute schedule task transaction: {:?}",
+        result
+    );
+
+    // Wait until the task scheduler actually mutates the account (with an upper bound to avoid hangs)
+    tokio::time::timeout(Duration::from_secs(1), async {
+        loop {
+            if env.get_account(account.pubkey()).data().first() == Some(&1) {
+                break;
+            }
+            tokio::time::sleep(Duration::from_millis(20)).await;
+        }
+    })
+    .await
+    .expect("task scheduler never incremented the account within 1s");
+
+    token.cancel();
+    handle.await.expect("task service join handle failed")?;
+
+    Ok(())
+}
+
+#[tokio::test]
+pub async fn test_cancel_task() -> TaskSchedulerResult<()> {
+    let (env, token, handle) = setup()?;
+
+    let account =
+        env.create_account_with_config(LAMPORTS_PER_SOL, 1, guinea::ID);
+
+    // Schedule a task
+    let task_id = 2;
+    let interval = 100;
+    let ix = Instruction::new_with_bincode(
+        guinea::ID,
+        &GuineaInstruction::ScheduleTask(ScheduleTaskArgs {
+            task_id,
+            execution_interval_millis: interval,
+            iterations: 100,
+            instructions: vec![Instruction::new_with_bincode(
+                guinea::ID,
+                &GuineaInstruction::Increment,
+                vec![AccountMeta::new(account.pubkey(), false)],
+            )],
+        }),
+        vec![
+            AccountMeta::new_readonly(magicblock_magic_program_api::ID, false),
+            AccountMeta::new(env.payer.pubkey(), true),
+            AccountMeta::new(account.pubkey(), false),
+        ],
+    );
+    let txn = env.build_transaction(&[ix]);
+    let result = env.execute_transaction(txn).await;
+    assert!(
+        result.is_ok(),
+        "failed to execute schedule task transaction: {:?}",
+        result
+    );
+
+    // Wait until we actually observe at least five executions
+    let executed_before_cancel =
```

Comment:
nit: in process executions are quite fast, so there's no need to wait that long, the subsequent loop with micro sleeps is more than enough.